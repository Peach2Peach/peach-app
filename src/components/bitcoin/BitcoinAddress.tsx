import Clipboard from '@react-native-clipboard/clipboard'
import { useRef } from 'react'
import { Animated, Pressable, TouchableOpacity, View } from 'react-native'
import QRCode from 'react-native-qrcode-svg'
import 'react-native-url-polyfill/auto'
import { IconType } from '../../assets/icons'
import { useIsMediumScreen } from '../../hooks/useIsMediumScreen'
import tw from '../../styles/tailwind'
import { getBitcoinAddressParts } from '../../utils/bitcoin/getBitcoinAddressParts'
import { openInWallet } from '../../utils/bitcoin/openInWallet'
import i18n from '../../utils/i18n'
import { Icon } from '../Icon'
import { PeachText } from '../text/PeachText'
import { BitcoinAddressProps } from './hooks/useBitcoinAddressSetup'

export const BitcoinAddress = ({ address, amount, label }: BitcoinAddressProps) => {
  const isMediumScreen = useIsMediumScreen()
  const width = isMediumScreen ? 327 : 242

  const requestTextOpacity = useRef(new Animated.Value(0)).current
  const addressTextOpacity = useRef(new Animated.Value(0)).current

  const urn = new URL(`bitcoin:${address}`)

  if (amount) urn.searchParams.set('amount', String(amount))
  if (label) urn.searchParams.set('message', label)

  const addressParts = getBitcoinAddressParts(address)

  const copyAddress = () => {
    Clipboard.setString(address)
    textAnimation(addressTextOpacity, 200)
  }

  const copyPaymentRequest = () => {
    Clipboard.setString(urn.toString())
    textAnimation(requestTextOpacity, 300)
  }

  const openInWalletOrCopyPaymentRequest = async () => {
    if (!(await openInWallet(urn.toString()))) copyPaymentRequest()
  }

  return (
    <>
      <Pressable onPress={openInWalletOrCopyPaymentRequest} onLongPress={copyPaymentRequest}>
        <Animated.View style={{ opacity: requestTextOpacity }}>
          <PeachText style={[tw`text-center subtitle-2`, tw`absolute w-20 -ml-10 bottom-full left-1/2`]}>
            {i18n('copied')}
          </PeachText>
        </Animated.View>
        <QRCode size={width} value={urn.toString()} backgroundColor={String(tw`text-primary-background`.color)} />
      </Pressable>

      <View style={tw`flex-row items-stretch w-full gap-2`}>
        <View style={tw`items-center justify-center px-3 py-2 border shrink border-black-25 rounded-xl`}>
          <PeachText style={tw`text-black-50`}>
            {addressParts.one}
            <PeachText style={tw`text-black-100`}>{addressParts.two}</PeachText>
            {addressParts.three}
            <PeachText style={tw`text-black-100`}>{addressParts.four}</PeachText>
          </PeachText>
          <Animated.View
            style={[
              tw`absolute items-center justify-center w-full h-full bg-primary-background-light`,
              { opacity: addressTextOpacity },
            ]}
          >
            <PeachText style={tw`text-center subtitle-1`}>{i18n('copied')}</PeachText>
          </Animated.View>
        </View>

        <View style={tw`justify-center gap-2`}>
          <IconButton onPress={copyAddress} iconId="copy" />
          <IconButton iconId="externalLink" onPress={openInWalletOrCopyPaymentRequest} />
        </View>
      </View>
    </>
  )
}

function IconButton ({ onPress, iconId }: { onPress: () => void; iconId: IconType }) {
  return (
    <TouchableOpacity onPress={onPress} style={tw`items-center px-4 py-1 bg-primary-main rounded-xl`}>
      <Icon id={iconId} size={24} color={tw.color('primary-background-light')} />
    </TouchableOpacity>
  )
}

function textAnimation (opacity: Animated.Value, duration: number) {
  Animated.sequence([
    Animated.timing(opacity, { toValue: 1, duration, useNativeDriver: true }),
    Animated.delay(1500),
    Animated.timing(opacity, { toValue: 0, duration, useNativeDriver: true }),
  ]).start()
}
