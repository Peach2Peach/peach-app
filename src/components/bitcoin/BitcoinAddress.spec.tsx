import ShallowRenderer from "react-test-renderer/shallow";
import { sellOffer } from "../../../tests/unit/data/offerData";
import { BitcoinAddress } from "./BitcoinAddress";

describe("BitcoinAddress", () => {
  const renderer = ShallowRenderer.createRenderer();

  it("should render correctly", () => {
    renderer.render(<BitcoinAddress address={sellOffer.returnAddress} />);

    const renderOutput = renderer.getRenderOutput();
    expect(renderOutput).toMatchSnapshot();
  });
  it("should render correctly for small screens", () => {
    const windowDimensionsSpy = jest.spyOn(
      jest.requireActual("react-native"),
      "useWindowDimensions",
    );
    windowDimensionsSpy.mockReturnValueOnce({ width: 320, height: 400 });
    renderer.render(<BitcoinAddress address={sellOffer.returnAddress} />);

    const renderOutput = renderer.getRenderOutput();
    expect(renderOutput).toMatchSnapshot();
  });
});
