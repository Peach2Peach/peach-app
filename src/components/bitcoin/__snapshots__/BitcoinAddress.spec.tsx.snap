// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitcoinAddress should render correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "column",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <Memo(Pressable)
    onLongPress={[Function]}
    onPress={[Function]}
  >
    <Fade
      delay={0}
      duration={300}
      show={false}
    >
      <PeachText
        style={
          [
            {
              "fontFamily": "Baloo2-SemiBold",
              "fontSize": 14,
              "letterSpacing": 0.14,
              "lineHeight": 22,
              "textAlign": "center",
            },
            {
              "bottom": "100%",
              "left": "50%",
              "marginLeft": -40,
              "position": "absolute",
              "width": 80,
            },
          ]
        }
      >
        copied!
      </PeachText>
    </Fade>
    <QRCode
      backgroundColor="#FFF9F6"
      size={327}
      value="bitcoin:bcrt1q70z7vw93cxs6jx7nav9cmcn5qvlv362qfudnqmz9fnk2hjvz5nus4c0fuh"
    />
  </Memo(Pressable)>
  <View
    style={
      {
        "alignItems": "stretch",
        "flexDirection": "row",
        "marginTop": 16,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#BAADA7",
          "borderRadius": 12,
          "borderWidth": 1,
          "flexShrink": 1,
          "justifyContent": "center",
          "marginRight": 8,
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
    >
      <Fade
        delay={0}
        duration={200}
        show={true}
        style={
          {
            "position": "absolute",
          }
        }
      >
        <PeachText
          style={
            {
              "color": "#9F8C82",
            }
          }
        >
          bcrt
          <PeachText
            style={
              {
                "color": "#2B1911",
              }
            }
          >
            1q70z
          </PeachText>
          7vw93cxs6jx7nav9cmcn5qvlv362qfudnqmz9fnk2hjvz5nus4
          <PeachText
            style={
              {
                "color": "#2B1911",
              }
            }
          >
            c0fuh
          </PeachText>
        </PeachText>
      </Fade>
      <Fade
        delay={0}
        duration={200}
        show={false}
        style={
          {
            "position": "absolute",
          }
        }
      >
        <PeachText
          style={
            {
              "fontFamily": "Baloo2-SemiBold",
              "fontSize": 16,
              "letterSpacing": 0.16,
              "lineHeight": 28,
              "textAlign": "center",
            }
          }
        >
          copied!
        </PeachText>
      </Fade>
    </View>
    <View>
      <PrimaryButton
        iconId="copy"
        onPress={[Function]}
      />
      <PrimaryButton
        iconId="externalLink"
        onPress={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`BitcoinAddress should render correctly for small screens 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "column",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <Memo(Pressable)
    onLongPress={[Function]}
    onPress={[Function]}
  >
    <Fade
      delay={0}
      duration={300}
      show={false}
    >
      <PeachText
        style={
          [
            {
              "fontFamily": "Baloo2-SemiBold",
              "fontSize": 14,
              "letterSpacing": 0.14,
              "lineHeight": 22,
              "textAlign": "center",
            },
            {
              "bottom": "100%",
              "left": "50%",
              "marginLeft": -40,
              "position": "absolute",
              "width": 80,
            },
          ]
        }
      >
        copied!
      </PeachText>
    </Fade>
    <QRCode
      backgroundColor="#FFF9F6"
      size={242}
      value="bitcoin:bcrt1q70z7vw93cxs6jx7nav9cmcn5qvlv362qfudnqmz9fnk2hjvz5nus4c0fuh"
    />
  </Memo(Pressable)>
  <View
    style={
      {
        "alignItems": "stretch",
        "flexDirection": "row",
        "marginTop": 16,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#BAADA7",
          "borderRadius": 12,
          "borderWidth": 1,
          "flexShrink": 1,
          "justifyContent": "center",
          "marginRight": 8,
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
    >
      <Fade
        delay={0}
        duration={200}
        show={false}
        style={
          {
            "position": "absolute",
          }
        }
      >
        <PeachText
          style={
            {
              "color": "#9F8C82",
            }
          }
        >
          bcrt
          <PeachText
            style={
              {
                "color": "#2B1911",
              }
            }
          >
            1q70z
          </PeachText>
          7vw93cxs6jx7nav9cmcn5qvlv362qfudnqmz9fnk2hjvz5nus4
          <PeachText
            style={
              {
                "color": "#2B1911",
              }
            }
          >
            c0fuh
          </PeachText>
        </PeachText>
      </Fade>
      <Fade
        delay={0}
        duration={200}
        show={
          {
            "fontScale": 2,
            "height": 1334,
            "scale": 2,
            "width": 750,
          }
        }
        style={
          {
            "position": "absolute",
          }
        }
      >
        <PeachText
          style={
            {
              "fontFamily": "Baloo2-SemiBold",
              "fontSize": 16,
              "letterSpacing": 0.16,
              "lineHeight": 28,
              "textAlign": "center",
            }
          }
        >
          copied!
        </PeachText>
      </Fade>
    </View>
    <View>
      <PrimaryButton
        iconId="copy"
        onPress={[Function]}
      />
      <PrimaryButton
        iconId="externalLink"
        onPress={[Function]}
        style={
          {
            "marginTop": 8,
          }
        }
      />
    </View>
  </View>
</View>
`;
